crates/hir/src/semantics.rs

```diff
@@ -173,6 +178,10 @@ impl<'db, DB: HirDatabase> Semantics<'db, DB> {
         self.imp.descend_node_at_offset(node, offset).find_map(N::cast)
     }
 
+    pub fn resolve_lifetime_param(&self, lifetime_token: &SyntaxToken) -> Option<LifetimeParam> {
```

Comment:
Hm, this definitely needs a `FIXME`, as this really should take an `ast::Something`. We can use a token or a node here. Tokens are specified in https://github.com/rust-analyzer/rust-analyzer/blob/8cba4231092b0d9ca44ea1f545856239e2fb6eed/xtask/src/codegen/gen_syntax.rs#L508-L511, nodes in the ungrammar repo. 

I suggest maybe starting with this untyped structure, and then refactor to `ast::` later, as this ordering should give us more context.